TestGOAWAYConcurrency fails with data race issues

This fails close to 100% of the time when run with race mode:

```
go test -race k8s.io/apiserver/pkg/server/filters -run TestGOAWAYConcurrency -v
```

Disabled in https://github.com/kubernetes/kubernetes/pull/94529

It is unclear if the test or the component needs to be corrected

/sig api-machinery
/priority critical-urgent

@answer1991 The test `TestClientReceivedGOAWAY/combine_with_watch-with-goaway` is also flaky.

https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/periodic-kubernetes-bazel-test-master/1302013439975100416

```
=== RUN   TestClientReceivedGOAWAY
=== RUN   TestClientReceivedGOAWAY/all_normal_requests_use_only_one_connection
=== RUN   TestClientReceivedGOAWAY/got_GOAWAY_after_set-up_watch
=== RUN   TestClientReceivedGOAWAY/got_GOAWAY_after_set-up_watch,_and_set-up_a_new_watch
=== RUN   TestClientReceivedGOAWAY/got_2_GOAWAY_after_set-up_watch
=== RUN   TestClientReceivedGOAWAY/combine_with_watch-with-goaway
    goaway_test.go:377: expect TCP connection: 3, actual: 4
--- FAIL: TestClientReceivedGOAWAY (20.11s)
```

but this one is hard to reproduce locally.

another case: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/94556/pull-kubernetes-bazel-test/1302300454738202625
```
=== RUN   TestClientReceivedGOAWAY/got_GOAWAY_after_set-up_watch,_and_set-up_a_new_watch
    goaway_test.go:377: expect TCP connection: 2, actual: 3
=== RUN   TestClientReceivedGOAWAY/got_2_GOAWAY_after_set-up_watch
=== RUN   TestClientReceivedGOAWAY/combine_with_watch-with-goaway
--- FAIL: TestClientReceivedGOAWAY (15.10s)
    --- PASS: TestClientReceivedGOAWAY/all_normal_requests_use_only_one_connection (0.01s)
    --- PASS: TestClientReceivedGOAWAY/got_GOAWAY_after_set-up_watch (5.01s)
    --- FAIL: TestClientReceivedGOAWAY/got_GOAWAY_after_set-up_watch,_and_set-up_a_new_watch (0.03s)
    --- PASS: TestClientReceivedGOAWAY/got_2_GOAWAY_after_set-up_watch (5.01s)
    --- PASS: TestClientReceivedGOAWAY/combine_with_watch-with-goaway (5.02s)
```

--- a/staging/src/k8s.io/apiserver/pkg/server/filters/goaway_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/server/filters/goaway_test.go
@@ -299,8 +299,6 @@ func requestGOAWAYServer(client *http.Client, serverBaseURL, url string) (<-chan
 // TestClientReceivedGOAWAY tests the in-flight watch requests will not be affected and new requests use a new
 // connection after client received GOAWAY.
 func TestClientReceivedGOAWAY(t *testing.T) {
-	t.Skip("disabled because of https://github.com/kubernetes/kubernetes/issues/94622")
-
 	s, err := newTestGOAWAYServer()
 	if err != nil {
 		t.Fatalf("failed to set-up test GOAWAY http server, err: %v", err)
@@ -438,10 +436,7 @@ func TestGOAWAYHTTP1Requests(t *testing.T) {
 }
 
 // TestGOAWAYConcurrency tests GOAWAY frame will not affect concurrency requests in a single http client instance.
-// Known issues in history: https://github.com/kubernetes/kubernetes/issues/91131.
 func TestGOAWAYConcurrency(t *testing.T) {
-	t.Skip("disabled because of https://github.com/kubernetes/kubernetes/issues/94532")
-
 	s, err := newTestGOAWAYServer()
 	if err != nil {
 		t.Fatalf("failed to set-up test GOAWAY http server, err: %v", err)

